### PR TITLE
WIP: Add python-accessible functions to capture vr controller events

### PR DESCRIPTION
**Pull request by drouin-simon**
and associate buttons with gestures and enable/disable interaction.

### CODE REVIEW

File: VirtualReality/MRML/vtkVirtualRealityViewInteractor.cxx [hunk 1]
```diff
@@ -162,3 +162,78 @@ void vtkVirtualRealityViewInteractor::SetTriggerButtonFunction(std::string funct
...
   }
 }
+
+//----------------------------------------------------------------------------
+void vtkVirtualRealityViewInteractor::SetGestureButtonToTrigger()
+{
+  vtkVirtualRealityViewInteractorStyle* vrInteractorStyle = vtkVirtualRealityViewInteractorStyle::SafeDownCast(this->InteractorStyle);
+  if (!vrInteractorStyle)
+  {
+    vtkWarningMacro("SetTriggerButtonFunction: Current interactor style is not a VR interactor style");
```
Comment (jcfr) [inline VirtualReality/MRML/vtkVirtualRealityViewInteractor.cxx:172]:
```suggestion
    vtkWarningMacro("SetGestureButtonToTrigger: Current interactor style is not a VR interactor style");
```

File: VirtualReality/MRML/vtkVirtualRealityViewInteractor.cxx [hunk 2]
```diff
@@ -162,3 +162,78 @@ void vtkVirtualRealityViewInteractor::SetTriggerButtonFunction(std::string funct
...
+  this->GestureEnabledButtons.push_back(static_cast<int>(vtkEventDataDeviceInput::Trigger));
+}
+
+//----------------------------------------------------------------------------
+void vtkVirtualRealityViewInteractor::SetGestureButtonToGrip()
+{
+  vtkVirtualRealityViewInteractorStyle* vrInteractorStyle = vtkVirtualRealityViewInteractorStyle::SafeDownCast(this->InteractorStyle);
+  if (!vrInteractorStyle)
+  {
+    vtkWarningMacro("SetTriggerButtonFunction: Current interactor style is not a VR interactor style");
```
Comment (jcfr) [inline VirtualReality/MRML/vtkVirtualRealityViewInteractor.cxx:190]:
```suggestion
    vtkWarningMacro("SetGestureButtonToGrip: Current interactor style is not a VR interactor style");
```

File: VirtualReality/MRML/vtkVirtualRealityViewInteractor.cxx [hunk 3]
```diff
@@ -162,3 +162,78 @@ void vtkVirtualRealityViewInteractor::SetTriggerButtonFunction(std::string funct
...
+  this->GestureEnabledButtons.push_back(static_cast<int>(vtkEventDataDeviceInput::Grip));
+}
+
+//----------------------------------------------------------------------------
+void vtkVirtualRealityViewInteractor::SetGestureButtonToTriggerAndGrip()
+{
+  vtkVirtualRealityViewInteractorStyle* vrInteractorStyle = vtkVirtualRealityViewInteractorStyle::SafeDownCast(this->InteractorStyle);
+  if (!vrInteractorStyle)
+  {
+    vtkWarningMacro("SetTriggerButtonFunction: Current interactor style is not a VR interactor style");
```
Comment (jcfr) [inline VirtualReality/MRML/vtkVirtualRealityViewInteractor.cxx:209]:
```suggestion
    vtkWarningMacro("SetGestureButtonToTriggerAndGrip: Current interactor style is not a VR interactor style");
```

File: VirtualReality/MRML/vtkVirtualRealityViewInteractor.cxx [hunk 4]
```diff
@@ -162,3 +162,78 @@ void vtkVirtualRealityViewInteractor::SetTriggerButtonFunction(std::string funct
...
+  this->GestureEnabledButtons.push_back(static_cast<int>(vtkEventDataDeviceInput::Trigger));
+}
+
+//----------------------------------------------------------------------------
+void vtkVirtualRealityViewInteractor::SetGestureButtonToNone()
+{
+  vtkVirtualRealityViewInteractorStyle* vrInteractorStyle = vtkVirtualRealityViewInteractorStyle::SafeDownCast(this->InteractorStyle);
+  if (!vrInteractorStyle)
+  {
+    vtkWarningMacro("SetTriggerButtonFunction: Current interactor style is not a VR interactor style");
```
Comment (jcfr) [inline VirtualReality/MRML/vtkVirtualRealityViewInteractor.cxx:229]:
```suggestion
    vtkWarningMacro("SetGestureButtonToNone: Current interactor style is not a VR interactor style");
```

**Comment by jcfr** [comment]
Topic has been rebased (along with replacing tabs with spaces, fixing minor issues, ...)

@cpinter When you have a chance could you review ?

File: VirtualReality/Widgets/qMRMLVirtualRealityView.h [hunk 5]
```diff
@@ -100,14 +100,51 @@ class Q_SLICER_QTMODULES_VIRTUALREALITY_WIDGETS_EXPORT qMRMLVirtualRealityView :
   /// Get underlying RenderWindow
   Q_INVOKABLE bool isHardwareConnected()const;
 
+  //--------------------------------------------------
+  // Hacks to limit interaction
```
Comment (jcfr) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.h:104]: Would it be possible to clarify this ?

**Comment by jcfr** [comment]
Before integrating, documentation  should probably updated to define `Trigger` vs `Grid` vs `TriggerAndGrip`

https://github.com/KitwareMedical/SlicerVirtualReality/blob/master/README.md
https://github.com/KitwareMedical/SlicerVirtualReality/blob/master/DeveloperGuide.md

**Comment by cpinter** [comment]
> When you have a chance could you review ?

It's high on my list. Hopefully tomorrow. I also want to keep working on the widget, but last week drew me completely away from projects/clients and have to catch up.

**Comment by jcfr** [comment]
Thanks for the follow up and great to know it is on your radar, no particular rush. :pray:

File: VirtualReality/Widgets/qMRMLVirtualRealityView.h (hunk 5 above)
Comment (cpinter) [reply VirtualReality/Widgets/qMRMLVirtualRealityView.h:104]: Maybe we could call these functions "Utility functions for convenient setting of frequently used configurations" ? @drouin-simon

**Comment by cpinter** [comment]
@lassoan Any major comments?

**Comment by lassoan** [comment]
It looks OK to me. When interaction with markups and other VTK MRML widgets will be implemented then we probably won't need these low-level event mappings anymore, but in the short term it is an effective workaround that enables prototyping of ideas.

**Comment by cpinter** [comment]
@drouin-simon I'm very sorry it took this long. I reviewed the code and made some fixes:
https://github.com/cpinter/SlicerVirtualReality/tree/controller-interaction

If there are no further comments I'll integrate this branch.

**Comment by cpinter** [comment]
Integrated in https://github.com/KitwareMedical/SlicerVirtualReality/pull/87

**Pull request closed without merging**